rfc/0000-define-matching-rules-for-form-urlencoded-body.md

```diff
@@ -149,6 +149,34 @@ pactffi_with_body(interaction, InteractionPart::Request, "application/x-www-form
 
 The values are ignored. Unsupported error messages are logged. The extracted example body will be empty.
 
+### Special cases
+
+- keys with no values (`a=&b=&c=`): can be achieved by define empty string value
+- values with no keys (`=a&=b&=c`): can't be achieved because empty key is not supported and will be ignored
+- repeated keys: (`a=1&a=2` or `a=2&a=1`): can be achieved by define array values
+- special characters
+    - `/`: will be encoded to `%2F`
+    - `&`: will be encoded to `%26`
+    - `?`: will be encoded to `%3F`
+    - `=`: will be encoded to `%3D`
+- only ampersands (`&&&`)
+    - can't be defined by json
+    - can be defined by raw syntax
+    - these query strings can match it:
+        - `&`
+        - `&&`
+        - `&&&`
+        - ...
+    - empty string can't match it
+- only equals signs (`===`)
+    - can't be defined by json
+    - can be defined by raw syntax
+    - these query strings can match it:
+        - `===`
+        - `=%3D%3D`
+    - empty string can't match it
+- no key and value (`=&=&=`): can't be achieved because empty key is not supported and will be ignored
+
```

Comment:
A couple of minor points here, and this is more in reference to the URL encoded RFC/specification I linked to above:

1. The `/` and `?` characters are typically _not_ special, even in a URL string. This means that an URL like `https://example.com/foo?a=?&?=/` has the query string `a=?&?=/`. This is an edge case which is often not properly handled, but it _is_ technically correct.
  - Having said that, to avoid confusion and possible implementation-specific variation, I would suggest that `/` and `%2F` both be allowed, and similarly with `?` and `%3F`. Effectively, we would just leave it up to the end-user to make the choice without Pact opining.
  - I agree that `&` and `=` still need to be escaped

For the `===` case, I would make it clear that its interpretation is ambiguous. I.e., there's no way to disambiguate parsing it as `{"=": "="}` or `{"": "=="}` or `{"==": ""}` (and even then, I'm not sure that is valid parsing... I should check the RFC again). I would specify that these should escape the `=` as above.

Lastly, you mention that empty string keys are not supported. Can you explain why? It would be ideal if we can support empty strings as we can then support `=a&=b&=c` as well as `=&=&=`.

Reply:
* `/` and `?` are currently encoded (both in key and value) in the current implementation of `process_json()` in pact-reference.
  * I think if we want to support these cases we need to change more code
  * I check with PHP's `http_build_query()` and it encoded them also i.e. `http_build_query(['/' => 'a', 'a' => '?', '?' => '/']) === '%2F=a&a=%3F&%3F=%2F'` 
* `===` can't be defined in json because theses cases are detected as plain text:
  * `{"=": "="}`
  * `{"": "=="}`
  * `{"==": ""}`
* Last time I tested empty key and it doesn't go well (matching rules fail always). Now I tested it again and it work fine, strange. So I support empty key again.